src/CoreWCF.Primitives/src/CoreWCF/Channels/ChannelDemuxer.cs

```diff
@@ -0,0 +1,3848 @@
+            //        {
+            //            DiagnosticUtility.TraceHandledException(e, TraceEventType.Information);
+            //        }
+            //        catch (TimeoutException e)
+            //        {
+            //            if (TD.SendTimeoutIsEnabled())
+            //            {
+            //                TD.SendTimeout(e.Message);
+            //            }
+            //            DiagnosticUtility.TraceHandledException(e, TraceEventType.Information);
+            //        }
+            //        catch (ObjectDisposedException e)
+            //        {
+            //            DiagnosticUtility.TraceHandledException(e, TraceEventType.Information);
+            //        }
+            //        catch (Exception e)
+            //        {
+            //            if (Fx.IsFatal(e)) throw;
+            //            this.HandleUnknownException(e);
+            //        }
+            //    }
+            //}
+            //finally
+            //{
+            //    if (abortItem)
+            //    {
+            //        this.AbortItem(request);
+            //    }
+            //}
+        }
+
+
+        protected override Message GetMessage(RequestContext request)
+        {
+            return request.RequestMessage;
+        }
+
+        internal class ReplyChannelDispatcher : IServiceChannelDispatcher
+        {
+            private ReplyChannelDemuxer demuxer;
+            private IChannel channel;
+
+            public ReplyChannelDispatcher(ReplyChannelDemuxer replyChannelDemuxer, IChannel channel)
+            {
+                this.demuxer = replyChannelDemuxer;
+                this.channel = channel;
+            }
+
+            public async Task DispatchAsync(RequestContext context)
+            {
+                // TODO: Find way to avoid instantiating a new ServiceChannelDispatcher each time
+                var serviceDispatcher = demuxer.MatchDispatcher(context.RequestMessage);
+                if (serviceDispatcher == null)
+                {
+                    CoreWCF.Dispatcher.ErrorBehavior.ThrowAndCatch(
+                        new EndpointNotFoundException(SR.Format(SR.UnableToDemuxChannel, context.RequestMessage.Headers.Action)), context.RequestMessage);
+                }
+                // TODO: if serviceDispatcher == null, use the EndpointNotFound code path
+                var serviceChannelDispatcher = await serviceDispatcher.CreateServiceChannelDispatcherAsync(channel);
+                await serviceChannelDispatcher.DispatchAsync(context);
+            }
+
+            public Task DispatchAsync(Message message)
+            {
+                throw new NotImplementedException();
+            }
+        }
+    }
+
+    interface IChannelDemuxerFilter
+    {
+        ChannelDemuxerFilter Filter { get; }
+    }
+
+//    class SingletonChannelListener<TChannel, TQueuedChannel, TQueuedItem> : DelegatingChannelListener<TChannel>, IChannelDemuxerFilter
```

Comment:
You can remove the SingletonChannelListener commend out class as it's no needed in the new model.